Dockerfile

```diff
@@ -1,24 +1,30 @@
 FROM paritytech/ci-linux:staging AS builder
+LABEL maintainer="imbue-dev"
 ARG PROFILE=release
-WORKDIR /imbue
-ADD . .
+ARG IMBUE_GIT_REPO="https://github.com/ImbueNetwork/imbue.git"
+ARG GIT_BRANCH="main"
+ARG GIT_CLONE_DEPTH="--depth 1"
+
+# WORKDIR /builds
+#Build imbue-collator
+RUN git clone --recursive ${IMBUE_GIT_REPO} ${GIT_CLONE_DEPTH}
+WORKDIR /builds/imbue
```

Comment:
I think pulling from git here disables some useful built-in functionality. For example, you can actually already [build an image from a git branch](https://docs.docker.com/engine/reference/commandline/build/#git-repositories):

```bash
docker build -t imbue-collator:custom https://github.com/ImbueNetwork/imbue.git#custom-branch
```

in fact, you can even specify which `Dockerfile`:

```bash
docker build -t imbue-collator:custom https://github.com/ImbueNetwork/imbue.git#custom-branch:scripts/Dockerfile.custom
```

And this is how we're using it in the ImbueNetwork/prototype repo's docker-compose.yaml:

```yaml
imbue:
  ...
  image: imbue-collator:latest
  build: http://github.com/ImbueNetwork/imbue.git#589ce100c1f8e76e3336920a7ce73653ae96d884
```

Now, if I try to use the functionality above in our docker-compose.yaml, it will do something confusing: it will go get the `Dockerfile` from that commit hash, but that `Dockerfile` will ignore the fact that I'm trying to use a specific commit, and instead builds from branch `main` unless I do other steps to work around that (that I'm not even sure is possible).

Also, I think the cache only invalidates because we're actually changing the content of files (e.g., the Dockerfile itself) -- that is, `ADD` is smart enough to know if it should invalidate cache or not, but we were not being smart enough about telling it which files to watch. We should probably use `ADD` to only add what's strictly necessary to do the build:

```Dockerfile
ADD libs libs
ADD node node
ADD pallets pallets
ADD runtime runtime
ADD Cargo.* ./
```

And then changes to any other file won't invalidate the cache.